### DISCUSSION TITLE
Programmatically set LIMIT in SPARQL query

### DISCUSSION THREAD

**Discussion by gdeak-monguz** [eclipse-rdf4j/rdf4j#5116] [Q&A] [answered]
### Current Behavior

I would like to limit the results of the query and I would like to pass the limit parameter as a variable. I use RDF4JTemplate from rdf4j-spring (and Kotlin). This is the query string:

```
const val SEARCH_QUERY = "PREFIX search: <http://www.openrdf.org/contrib/lucenesail#> " +
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#> " +
      "SELECT ?subject ?type ?label ?highlightedLabel ?limit " +
      "WHERE { " +
      "?subject a ?type ; " +
      "rdfs:label ?label . " +
      "?subject search:matches [ " +
      "search:query ?searchTerm ; " +
      "search:property rdfs:label ; " +
      "search:snippet ?highlightedLabel ] " +
      "} " +
      "ORDER BY ?label " +
      "LIMIT ?limit"
```

And this is the code that run this query:
```
template.tupleQuery(SEARCH_QUERY)
      .withBinding(SEARCH_TERM, "test*")
      .withBinding(LIMIT, 10)
      .evaluateAndConvert()
```

I got this error:

```
org.eclipse.rdf4j.query.MalformedQueryException: Encountered " <VAR1> "?limit "" at line 1, column 356.
Was expecting:
    <INTEGER> ...
    ] with root cause

org.eclipse.rdf4j.query.parser.sparql.ast.ParseException: Encountered " <VAR1> "?limit "" at line 1, column 356.
Was expecting:
    <INTEGER> ...
    ```

### Expected Behavior

Set limit with the `withBindig`method.

### Version

5.0.2

### Are you interested in contributing a solution yourself?

None

**Comment by fkleedorfer** [comment]
As per the specification, limit only accepts an integer literal: https://www.w3.org/TR/sparql11-query/#rLimitClause

**Comment by fkleedorfer** [comment]
... so I don't believe it's a bug, or is it?

**Comment by gdeak-monguz** [comment]
> ... so I don't believe it's a bug, or is it?

@fkleedorfer with SparqlBuilder it is possible to specify the limit with the `.limit()` function of the builder, but as you can see I would like to use Lucene search features that is not supported by the builder. I think it is a bug, but I am not sure. Maybe it is not a bug, but a defect of RDF.

**Comment by fkleedorfer** [comment, marked as answer]
This should work. It's not pretty, though. We may want to add a way to add anonymous blank nodes in a triple pattern.

```
    public static void main(String[] args) {
        Variable subject = SparqlBuilder.var("subject");
        Variable type = SparqlBuilder.var("type");
        Variable label = SparqlBuilder.var("label");
        Variable highlightedLabel = SparqlBuilder.var("highlightedLabel");
        Variable searchTerm = SparqlBuilder.var("searchTerm");
        Prefix search = SparqlBuilder.prefix("search", Rdf.iri("http://www.openrdf.org/contrib/lucenesail#"));
        int limit = 20;
        RdfBlankNode searchSpec = Rdf.bNode("searchSpec");
        SelectQuery query = Queries
                        .SELECT(subject, type, label, highlightedLabel)
                        .prefix(search)
                        .prefix(RDFS.NS)
                        .where(
                                        subject
                                                        .isA(type)
                                                        .andHas(RDFS.LABEL, label)
                                                        .andHas( search.iri("matches"), searchSpec)
                                                        .and(searchSpec
                                                                                        .has(search.iri("query"), searchTerm)
                                                                                        .andHas(search.iri("property"), RDFS.LABEL)
                                                                                        .andHas(search.iri("snippet"), highlightedLabel)))
                        .orderBy(label)
                        .limit(limit);
        System.out.println(query.getQueryString());
    }
```
this prints
```
PREFIX search: <http://www.openrdf.org/contrib/lucenesail#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?subject ?type ?label ?highlightedLabel
WHERE { ?subject a ?type ;
    rdfs:label ?label ;
    search:matches _:searchSpec .
_:searchSpec search:query ?searchTerm ;
    search:property rdfs:label ;
    search:snippet ?highlightedLabel . }
ORDER BY ?label
LIMIT 20
```

**Comment by hmottestad** [comment]
The purpose of binding variables at query time is primarily to inject unsafe values and mitigate attacks equivalent to SQL-injection.

For values you know are safe you can always revert to doing a simple search/replace on the query string. Just use a unique variable name such as one based off a UUID.

**Comment by hmottestad** [comment]
We can rename this issue to be about the SparqlBuilder if that makes more sense.

**Comment by fkleedorfer** [comment]
Apart from its developer experience being quite puzzling at times, the problem here is that the SparqlBuilder does not allow you to specify anonymous blank nodes with predicate-object lists. That would be a relatively simple addition, I believe.

**Comment by fkleedorfer** [comment]
Just to make sure your problem is solved @gdeak-monguz, did [this SparqlBuilder solution](https://github.com/eclipse-rdf4j/rdf4j/issues/5111#issuecomment-2304482707) work for you ?

**Comment by gdeak-monguz** [comment]
> Just to make sure your problem is solved @gdeak-monguz, did [this SparqlBuilder solution](https://github.com/eclipse-rdf4j/rdf4j/issues/5111#issuecomment-2304482707) work for you ?

Yes, thank you for your help!
